torch_xla/csrc/xla_sharding_util.cpp

```diff
@@ -641,10 +639,46 @@ ShardingUtil::GetAutoShardingMesh() {
                  runtime::GetComputationClient()->GetAllDevices().size())
         << "Invalid auto-sharding mesh_shape: "
         << absl::StrJoin(mesh_shape, ",");
-    device_mesh_ids = std::vector<int64_t>(total_devices);
-    std::iota(device_mesh_ids.begin(), device_mesh_ids.end(), 0);
   }
-  return std::make_tuple(mesh_shape, device_mesh_ids);
+  return mesh_shape;
+}
+
+std::vector<int64_t> ShardingUtil::GetAutoShardingMeshIds(
+    const xla::HloModuleProto& module) {
+  std::vector<xla::OpSharding> input_shardings;
+  // Unforuntately, we have to go through the instructions since
+  // `spmd_parameters_shardings` is not available.
+  for (auto computation : module.computations()) {
+    for (auto instruction : computation.instructions()) {
+      if (instruction.opcode() == "parameter" && instruction.has_sharding()) {
+        input_shardings.push_back(instruction.sharding());
+      }
+    }
+  }
+
+  // Return the first non-default (iota) mesh ids arrangement, as we expect
+  // only one such assignment and/or the logical mesh device assignment should
+  // be compatible with the other arrangements in the HLO. This is a work-around
+  // as the auto-sharding pass takes only one arrangement for now.
+  // TODO(yeounoh) this was not necessary before; replace if this can be done
+  // during the auto-sharding pass.
+  int64_t n_devices = runtime::GetComputationClient()->GetAllDevices().size();
+  std::vector<int64_t> device_mesh_ids = std::vector<int64_t>(n_devices);
+  std::iota(device_mesh_ids.begin(), device_mesh_ids.end(), 0);
+  for (auto sharding : input_shardings) {
+    std::vector<int64_t> new_mesh_ids;
+    auto tile_assignment_devices = sharding.tile_assignment_devices();
+    if (!tile_assignment_devices.empty()) {
+      new_mesh_ids = std::vector<int64_t>(tile_assignment_devices.begin(),
+                                          tile_assignment_devices.end());
+      // return the first non-default (iota) device assigments.
+      if (new_mesh_ids != device_mesh_ids) {
+        return new_mesh_ids;
+      }
+    }
```

Comment:
nit: instead of creating the `input_shardings` array, can you compute new_mesh_id as you loop through the instruction? This way you won't waste time to get the inpur_shardings you don't need.

Reply:
Ack

Reply:
done